draft-rfcs/const-evaluatable-checked.md

```diff
@@ -0,0 +1,19 @@
+# const evaluatable checked
+
+TODO: All of this :)
+
+## Requirements
+
+For this to work, const operations have to be deterministic and
+must not depend on any external state,
+at least when they are used in the type system.
+
+Using floats during CTFE is fully determinstic. So using
+them inside of the type system is fine. CTFE can however
+produce different results than what would happen on real hardware,
+but this is not a concern for const generics.
+
+Other sources of non-determinism are allocations. This non-determinism
+must however not be observed during const-evaluation (TODO: link to const-eval)
+and is only an issue if allocations leaked into the result. This
```

Comment:
Valtrees are not about "not leaking allocations to the result". The point of a valtree is that a value of type `&i32` is identified only by the value it points to, not the location in memory. So two `&i32` are considered equal if they point to equal things, even if these equal things are located at different addresses.

This is what `PartialEq` does anyway, but in the compiler, the result of a CTFE query is typically represented as an `Allocation` and comparing those will also compare the underlying (abstract) addresses.

Reply:
Ah yeah, messed this up :+1: Was thinking about how we deal with mutable references which isn't too related to val trees.

Thinking about this, how do we intend to deal with the following:
```rust
#![feature(const_generics)]
fn foo<const N: &'static i32>() -> usize {
    N as *const i32 as usize
}

fn bar<const M: &'static i32>() {
    assert_eq!(M as *const i32 as usize, foo::<M>());
}

fn main() {
    foo::<{ &3 }>();
    bar::<{ &3 }>(); 
}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=79e11d1a3938217abd4171234bf6d782)
Here we have two different anonymous constants which potentially create a different allocation for `3`.
These do result in the same valtrees though and should be considered equal by the type system.

This means that during monomorphization we should end up with only one instance for `foo`. I think that we want to guarantee that this program does not panic, but how do we assert that we only have one unique allocation for all references to `3` used in types? Especially if we codegen `foo` in different crates.

Reply:
> how do we assert that we only have one unique allocation for all references to 3 used in types

We don't, and we don't need to.

The point of valtrees is that we convert the CTFE result to a valtree, and use valtree equality tests only in the type system. So it won't matter if the two allocations with `3` in them have the same address or not; the result valtree will be the same either way.

Reply:
>  I want to guarantee that this program does not panic

Oh, I missed that part. No that won't be possible I'm afraid. If programmers "look behind the curtain" of the notion of equality we use, they have to deal with the consequences... we currently make no guarantees whatsoever for address identity in `const`, and I don't see what guarantees we *could* provide.